src/common/model/set/setParameters.ml

```diff
@@ -1,10 +1,17 @@
 open Nes
 
+(** How to render the order. [Default] prints the tunes as they appear in the
+    set. [Unfolded] follows the order, duplicating the tunes if they are to be
+    played several times. *)
+type order_type = Default | Unfolded
+[@@deriving yojson]
+
 module Self = struct
   type t =
     { forced_pages  : int    option [@default None] [@key "forced-pages"] ;
       show_deviser  : bool   option [@default None] [@key "show-deviser"] ;
       show_order    : bool   option [@default None] [@key "show-order"] ;
+      order_type    : order_type option [@default None] [@key "order-type"] ;
```

Comment:
Since Default already models the "no additional information" case, is it redundant with None? Should this just be an `order_type`?

Reply:
First, there are two notions of “default” here: one is the notion of default for the `make` function and the JSON serialisation: what should this field be when it is not present in the JSON or in the arguments of `make`; the second one is the notion of default value in Dancelor, eg. when we want to generate a PDF.

Now I don't find this very satisfactory but basically those things are parameter _diffs_ and not really parameters. In particular, we want to be able to represent diffs that do not define the order type, relying on the inherited value. Those would be `None`. So we do need those things to be `option`s and we do need the `[@default None]` so that Yojson (in particular) knows to put `None` there when the field does not appear in the serialisation.

The whole thing is not very satisfactory because there is then no guarantee that the values are set at all when you're holding “parameters”. This is why some fields getters have an `Option.unwrap` in front of them, but that is quite nasty and I don't like it.

Also, we use two PPXes (`make` and `yojson`) that both interpret the `[@default]` differently. For `yojson`, it means that you should use that if field is not present in the JSON; fairly straightforward. `make`, however, attempts to be clever and already considers that if it's an `option` then it should be an optional parameter. By giving it a `[@default]`, though, we make it doubly optional, so `make` would have type `?order_type: order_type option -> ...` which is an optional argument that holds an option. This is why we have those ugly `let make ?order_type ... = make ~order_type` things: to get rid of one level of option-ness.

Does that make the situation a bit clearer? And answers your question?

Reply:
It does! Thanks for the explanation.